### PR TITLE
VMHostVss

### PR DESCRIPTION
**Pull request by lucdekens**
- VMHostVss resource only contains the basic VSS settings
- Unit test script without hard-coded values
- Unit test in Try-Finally construct

Signed-off-by: Luc Dekens <dekens.luc@gmail.com>

### CODE REVIEW

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+        Value = $script:Constants.NetworkSystemMoRefValue
+    }
+}
+
+function AfterAllTests {
+    Remove-Module -Name VMware.VimAutomation.Core
+    $env:PSModulePath = $script:modulePath
+}
+
+# Calls the function to Import the mocked VMware.VimAutomation.Core module before all tests.
```
Comment (SimeonGerginov) [inline Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Move the comment in the Try clause.

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+Try {
+    BeforeAllTests
+
+    Describe 'VMHostVss\Set'  -Tag 'Set' {
+        Context 'Present + VSS does not exist' {
```
Comment (SimeonGerginov) [inline Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: For better readability change the plus to 'and'.

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+            Mock -CommandName Update-Network -MockWith $updateNetworkMock -ModuleName $script:moduleName
+
+            $resource = New-Object -TypeName $script:resourceName -Property $script:resourceProperties
+
+            It 'Should call the Connect-VIServer mock with the passed server and credentials once' {
+                # Act
+                $resource.Set()
+
+                # Assert
+                $sAssert = @{
```
Comment (SimeonGerginov) [inline Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Maybe here we should think about better naming.

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+            $result.Server | Should -Be $script:resourceProperties.Server
+            $result.Credential | Should -Be $script:resourceProperties.Credential
+            $result.Name | Should -Be $script:resourceProperties.Name
+            $result.Ensure | Should -Be $script:resourceProperties.Ensure
+            $result.VssName | Should -Be $script:resourceProperties.VssName
+            $result.NumPorts | Should -Be $script:resourceProperties.NumPorts
+            $result.Mtu | Should -Be $script:resourceProperties.Mtu
+            $result.Key | Should -Be $script:resourceProperties.Key
+            $result.NumPortsAvailable | Should -Be $script:resourceProperties.NumPortsAvailable
+            [string]::Compare(($result.Pnic | ConvertTo-Json), ($script:resourceProperties.Pnic | ConvertTo-Json)) | Should -Be 0
```
Comment (SimeonGerginov) [inline Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Here you can try the Compare-Object function that is part of the repository. Example usage:
```powershell
 Compare-Object -ReferenceObject $SatpArgs.Values -DifferenceObject $addSatpArgs.Values
```

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 3 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: You mean the name of the splat variable?

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 4 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Compare-Object is a standard PS cmdlet, but i'll change it

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 3 above)
Comment (SimeonGerginov) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Yes.

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 4 above)
Comment (SimeonGerginov) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: I think you can use it without changing it:
```powershell
 -ParameterFilter { $EsxCli.Id -eq $esxcliObject.Id -and ( Compare-Object -ReferenceObject $SatpArgs.Values -DifferenceObject $addSatpArgs.Values ) -eq $null }
```

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 3 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Any suggestions?
That's my personal convention, 's' for splat variable and then the verb of the cmdlet where I use it.
And it is a local variable for the It block.

**Comment by lucdekens** [comment]
Except for the names of the splat variables all requested changes should be in.

File: Documentation/DSCResources/VMHostVss/VMHostVss.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
+# VMHostVss
+
+## Parameters
+
+| Parameter | Attribute | DataType | Description | Allowed Values |
+| --- | --- | --- | --- | --- |
+| **Server** | Key | string | Name of the Server we are trying to connect to. The Server can be a vCenter or ESXi. ||
+| **Name** | Key | string | Name of the VMHost to configure. ||
+| **Credential** | Mandatory | PSCredential | Credentials needed for connection to the specified Server. ||
+| **Ensure** | Mandatory | Ensure | Value indicating if the VSS should be Present or Absent. ||
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Documentation/DSCResources/VMHostVss/VMHostVss.md]: Done

File: Documentation/DSCResources/VMHostVss/VMHostVss.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+## Description
+
+The resource is used to configure the basic properties of a Virtual Switch (VSS) on an ESXi node.
+
+## Examples
+
+### Example 1
+
+Create a new VSS on an ESXi node.
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Documentation/DSCResources/VMHostVss/VMHostVss.md]: Done

File: Documentation/DSCResources/VMHostVss/VMHostVss.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+Configuration VMHostVss_Config {
+    Import-DscResource -ModuleName VMware.vSphereDSC
+
+    Node localhost {
+        $Password = $Password | ConvertTo-SecureString -AsPlainText -Force
+        $Credential = New-Object System.Management.Automation.PSCredential($User, $Password)
+
+        VMHostVss vmHostVSS
+        {
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Documentation/DSCResources/VMHostVss/VMHostVss.md]: Done

File: Documentation/DSCResources/VMHostVss/VMHostVss.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            Mtu = 1500
+            NumPorts = 1
+        }
+    }
+}
+````
+
+### Example 2
+
+Remove a VSS from an ESXi node.
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Documentation/DSCResources/VMHostVss/VMHostVss.md]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            return ($null -eq $vss)
+        }
+    }
+
+    [VMHostVss] Get() {
+        Write-Verbose -Message "$(Get-Date) $($s = Get-PSCallStack; "Entering {0}" -f $s[0].FunctionName)"
+
+        $result = [VMHostVss]::new()
+        $result.Server = $this.Server
+        $result.Credential = $this.Credential
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        $result.Name = $this.Name
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        $vmHost = $this.GetVMHost()
+        $this.PopulateResult($vmHost, $result)
+
+        return $result
+    }
+
+    <#
+    .DESCRIPTION
+
+    Returns a boolean value indicating if the VMHostVss should to be updated.
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+        return ($vssTest -notcontains $false)
+    }
+
+    <#
+    .DESCRIPTION
+
+    Returns the desired virtual switch if it is present on the server otherwise returns $null.
+    #>
+    [PSObject]GetVss($vmHost) {
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        Write-Verbose -Message "$(Get-Date) $($s = Get-PSCallStack; "Entering {0}" -f $s[0].FunctionName)"
+
+        $hostNetSys = Get-View -Server $this.Connection -Id $vmHost.ExtensionData.ConfigManager.NetworkSystem
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        $vss = $hostNetSys.NetworkInfo.Vswitch |
+            Where-Object {$_.Name -eq $this.VssName}
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    Updates the configuration of the virtual switch.
+    #>
+    [void] UpdateVss($vmHost) {
+        Write-Verbose -Message "$(Get-Date) $($s = Get-PSCallStack; "Entering {0}" -f $s[0].FunctionName)"
+
+        $vssConfigArgs = @{
+            Name = $this.VssName
+            Mtu = $this.Mtu
+            NumPorts = $this.NumPorts
+            Operation = 'add'
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        $vss = $this.GetVss($vmHost)
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        else {
+            if ($null -eq $vss) {
+                return
+            }
+            $vssConfigArgs.Operation = 'remove'
+        }
+
+        $vssConfig = New-VssConfig @vssConfigArgs
+        $hostNetSys = Get-View -Server $this.Connection -Id $vmHost.ExtensionData.ConfigManager.NetworkSystem
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+        try {
+            Update-Network -NetworkSystem $hostNetSys -Type 'VSS' -VssConfig $vssConfig
+        }
+        catch {
+            Write-Error "The virtual switch Config could not be updated: $($PSItem.ToString())"
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        return
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+
+    <#
+    .DESCRIPTION
+
+    Populates the result returned from the Get() method with the values of the virtual switch.
+    #>
+    [void] PopulateResult($vmHost, $vmHostVSS) {
+        Write-Verbose -Message "$(Get-Date) $($s = Get-PSCallStack; "Entering {0}" -f $s[0].FunctionName)"
+
+        $currentVss = $this.GetVss($vmHost)
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/Tests/Integration/VMHostVss/VMHostVss.Integration.Tests.ps1 [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+$script:dscResourceName = 'VMHostVss'
+$script:dscConfig = $null
+$script:moduleFolderPath = (Get-Module VMware.vSphereDSC -ListAvailable).ModuleBase
+$script:integrationTestsFolderPath = Join-Path (Join-Path (Join-Path $moduleFolderPath 'Tests') 'Integration') $script:dscResourceName
+
+$script:configWithNewVss = "$($script:dscResourceName)_New_Config"
+$script:configWithModifyVss = "$($script:dscResourceName)_Modify_Config"
+$script:configWithRemoveVss = "$($script:dscResourceName)_Remove_Config"
+
+$script:connection = Connect-VIServer -Server $Server -User $User -Password $Password
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Integration/VMHostVss/VMHostVss.Integration.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/Tests/Integration/VMHostVss/VMHostVss.Integration.Tests.ps1 [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+        It 'Should be able to call Get-DscConfiguration without throwing' {
+            # Assert
+            { Get-DscConfiguration } | Should -Not -Throw
+        }
+
+        It 'Should be able to call Get-DscConfiguration and all the parameters should match' {
+            # Act
+            $configuration = Get-DscConfiguration
+
+            # Assert
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Integration/VMHostVss/VMHostVss.Integration.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/Tests/Unit/TestHelpers/VMware.VimAutomation.Core/VMwareVimTypes.cs [hunk 23]
```diff
@@ -643,4 +647,183 @@ public override int GetHashCode()
...
+
+        public int NumPortsAvailable { get; set; }
+
+        public string[] Pnic { get; set; }
+
+        public string[] Portgroup { get; set; }
+
+        public bool Equals(HostVirtualSwitch hostVirtualSwitch)
+        {
+            return hostVirtualSwitch != null && this.Name == hostVirtualSwitch.Name && this.NumPorts == hostVirtualSwitch.NumPorts &&
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/TestHelpers/VMware.VimAutomation.Core/VMwareVimTypes.cs]: Done

File: Source/VMware.vSphereDSC/Tests/Unit/TestHelpers/VMware.VimAutomation.Core/VMwareVimTypes.cs [hunk 24]
```diff
@@ -643,4 +647,183 @@ public override int GetHashCode()
...
+        public override int GetHashCode()
+        {
+            return (this.Key + "_" +
+                    this.Mtu + "_" +
+                    this.Name + "_" +
+                    this.NumPorts + "_").GetHashCode();
+        }
+    }
+
+    public class HostVirtualSwitchBridge
```
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/TestHelpers/VMware.VimAutomation.Core/VMwareVimTypes.cs]: Done

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 1 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 2 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 4 above)
Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 (hunk 9 above)
[the remark this answers is not in the document]
Comment (SimeonGerginov) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Can you apply little ordering here for better readability:
```powershell
 $this.ConnectVIServer()
 $vmHost = $this.GetVMHost()

 $result = [VMHostVss]::new()
 $result.Server = $this.Server
 $result.Name = $vmHost.Name

 $this.PopulateResult($vmHost, $result)

 if ($null -ne $result.Key) {
     $result.Ensure = 'Present'
 }
 else {
     $result.Ensure = 'Absent'
 }

 return $result
```

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 (hunk 14 above)
[the remark this answers is not in the document]
Comment (SimeonGerginov) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: And if you can add spaces like this: 
```powershell
 Where-Object { $_.Name -eq $this.VssName }
```

File: Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1 (hunk 3 above)
Comment (SimeonGerginov) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: I was thinking of something like $assertMockCalledParams.

Comment (lucdekens) [reply Source/VMware.vSphereDSC/Tests/Unit/VMHostVss.Unit.Tests.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 (hunk 14 above)
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

File: Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1 (hunk 9 above)
[the remark this answers is not in the document]
Comment (lucdekens) [reply Source/VMware.vSphereDSC/DSCResources/VMHostVss/VMHostVss.ps1]: Done

**Pull request merged**